### PR TITLE
[FEATURE] Publish stable types for Ember

### LINKED ISSUE #19946: [QUEST] Future Public TypeScript Types
The plan is eventually for Ember to export all of its own TypeScript types. In preparation for this, we'd like to have the internal Ember modules export the correct types.

## RFC 0176 Types

These are the full set of APIs Ember has exposed, per [the officially maintained data](https://github.com/ember-cli/ember-rfc176-data). This may include items that have been deprecated or removed!

<details><summary><b>Detailed Status breakdown</b></summary>

Unchecked items are one of the following:
* Unimplemented
* Untested
* Things we don't want to export

### `@ember/application`
* [x] `import Application from '@ember/application':` - `Ember.Application`
* [x] `import { getOwner } from '@ember/application';` - `Ember.getOwner`
* [ ] ~`import { onLoad } from '@ember/application';` - `Ember.onLoad`~
* [ ] ~`import { runLoadHooks } from '@ember/application';` - `Ember.runLoadHooks`~
* [x] `import { setOwner } from '@ember/application';` - `Ember.setOwner`
* [ ] ~`import GlobalsResolver from '@ember/application/globals-resolver';` - `Ember.DefaultResolver`~
* [x] `import ApplicationInstance from '@ember/application/instance';` - `Ember.ApplicationInstance`
* [x] `import Namespace from '@ember/application/namespace';` - `Ember.Namespace`
* [ ] ~`import Resolver from '@ember/application/resolver';` - `Ember.Resolver`~

### `@ember/array`
* [x] `import EmberArray from '@ember/array';` - `Ember.Array`
* [x] `import { A } from '@ember/array';` - `Ember.A`
* [x] `import { isArray } from '@ember/array';` - `Ember.isArray`
* [x] `import { makeArray } from '@ember/array';` - `Ember.makeArray`
* [x] `import MutableArray from '@ember/array/mutable';` - `Ember.MutableArray`
* [x] `import ArrayProxy from '@ember/array/proxy';` - `Ember.ArrayProxy`

### `@ember/canary-features` (PR: #19947)
* [x] `import { FEATURES } from '@ember/canary-features';` - `Ember.FEATURES`
* [x] `import { isEnabled } from '@ember/canary-features';` - `Ember.FEATURES.isEnabled`

### `@ember/component` (PR: #19948)
* [x] `import Component from '@ember/component';` - `Ember.Component`
* [x] `import { Input } from '@ember/component';` - `Ember._Input`
* [x] `import { Textarea } from '@ember/component';` - `Ember._TextArea`
* [x] `import { capabilities } from '@ember/component';` - `Ember._componentManagerCapabilities`
* [x] **UNTESTED** `import { getComponentTemplate } from '@ember/component';` - `Ember._getComponentTemplate`
* [x] **UNTESTED** `import { setComponentManager } from '@ember/component';` - `Ember._setComponentManager`
* [x] `import { setComponentTemplate } from '@ember/component';` - `Ember._setComponentTemplate`
* [ ] ~`import Checkbox from '@ember/component/checkbox';` - `Ember.Checkbox`~
* [x] `import Helper from '@ember/component/helper';` - `Ember.Helper`
* [x] `import { helper as buildHelper } from '@ember/component/helper';` - `Ember.Helper.helper`
* [x] `import templateOnlyComponent from '@ember/component/template-only';` - `Ember._templateOnlyComponent`
* [ ] ~`import TextArea from '@ember/component/text-area';` - `Ember.TextArea`~
* [ ] ~`import TextField from '@ember/component/text-field';` - `Ember.TextField`~

### `@ember/controller`
* [x] `import Controller from '@ember/controller';` - `Ember.Controller`
* [x] `import { inject } from '@ember/controller';` - `Ember.inject.controller`

### `@ember/debug` (PR: #19947)
* [x] `import { assert } from '@ember/debug';` - `Ember.assert`
* [x] `import { debug } from '@ember/debug';` - `Ember.debug`
* [ ] ~`import { inspect } from '@ember/debug';` - `Ember.inspect`~
* [x] `import { registerDeprecationHandler } from '@ember/debug';` - `Ember.Debug.registerDeprecationHandler`
* [x] `import { registerWarnHandler } from '@ember/debug';` - `Ember.Debug.registerWarnHandler`
* [x] `import { runInDebug } from '@ember/debug';` - `Ember.runInDebug`
* [x] `import { warn } from '@ember/debug';` - `Ember.warn`
* [x] `import ContainerDebugAdapter from '@ember/debug/container-debug-adapter';` - `Ember.ContainerDebugAdapter`
* [x] `import DataAdapter from '@ember/debug/data-adapter';` - `Ember.DataAdapter`

### `@ember/destroyable` (PR: #19947)
* [x] `import { assertDestroyablesDestroyed } from '@ember/destroyable';` - `Ember._assertDestroyablesDestroyed`
* [x] `import { associateDestroyableChild } from '@ember/destroyable';` - `Ember._associateDestroyableChild`
* [x] `import { destroy } from '@ember/destroyable';` - `Ember.destroy`
* [x] `import { enableDestroyableTracking } from '@ember/destroyable';` - `Ember._enableDestroyableTracking`
* [x] `import { isDestroyed } from '@ember/destroyable';` - `Ember._isDestroyed`
* [x] `import { isDestroying } from '@ember/destroyable';` - `Ember._isDestroying`
* [x] `import { registerDestructor } from '@ember/destroyable';` - `Ember._registerDestructor`
* [x] `import { unregisterDestructor } from '@ember/destroyable';` - `Ember._unregisterDestructor`

### `@ember/engine` (PR: #19923)
* [x] `import Engine from '@ember/engine';` - `Ember.Engine`
* [x] `import { getEngineParent } from '@ember/engine';` - `Ember.getEngineParent`
* [x] `import EngineInstance from '@ember/engine/instance';` - `Ember.EngineInstance`

### `@ember/enumerable`
* [x] `import Enumerable from '@ember/enumerable';` - `Ember.Enumerable`

### `@ember/error` (PR: #19947)
* [x] `import EmberError from '@ember/error';` - `Ember.Error`

### `@ember/helper` (PR: #19947)
* [x] `import { array } from '@ember/helper';` - `Ember._array`
* [x] `import { capabilities } from '@ember/helper';` - `Ember._helperManagerCapabilities`
* [x] `import { concat } from '@ember/helper';` - `Ember._concat`
* [x] `import { fn } from '@ember/helper';` - `Ember._fn`
* [x] `import { get } from '@ember/helper';` - `Ember._get`
* [x] `import { hash } from '@ember/helper';` - `Ember._hash`
* [x] `import { invokeHelper } from '@ember/helper';` - `Ember._invokeHelper`
* [x] `import { setHelperManager } from '@ember/helper';` - `Ember._setHelperManager`

### `@ember/instrumentation`
* [ ] `import { instrument } from '@ember/instrumentation';` - `Ember.Instrumentation.instrument`
* [ ] `import { reset } from '@ember/instrumentation';` - `Ember.Instrumentation.reset`
* [ ] `import { subscribe } from '@ember/instrumentation';` - `Ember.Instrumentation.subscribe`
* [ ] `import { unsubscribe } from '@ember/instrumentation';` - `Ember.Instrumentation.unsubscribe`

### `@ember/map`
* [ ] `import EmberMap from '@ember/map';` - `Ember.Map`
* [ ] `import MapWithDefault from '@ember/map/with-default';` - `Ember.MapWithDefault`

### `@ember/modifier`
* [x] `import { capabilities } from '@ember/modifier';` - `Ember._modifierManagerCapabilities`
* [x] `import { on } from '@ember/modifier';` - `Ember._on`
* [x] `import { setModifierManager } from '@ember/modifier';` - `Ember._setModifierManager`

### `@ember/object`
* [x] `import EmberObject from '@ember/object';` - `Ember.Object`
* [x] `import { action } from '@ember/object';` - `Ember._action`
* [ ] ~`import { aliasMethod } from '@ember/object';` - `Ember.aliasMethod`~
* [x] `import { computed } from '@ember/object';` - `Ember.computed`
* [x] `import { defineProperty } from '@ember/object';` - `Ember.defineProperty`
* [x] `import { get } from '@ember/object';` - `Ember.get`
* [x] `import { getProperties } from '@ember/object';` - `Ember.getProperties`
* [ ] ~`import { getWithDefault } from '@ember/object';` - `Ember.getWithDefault`~
* [x] `import { notifyPropertyChange } from '@ember/object';` - `Ember.notifyPropertyChange`
* [x] `import { observer } from '@ember/object';` - `Ember.observer`
* [x] `import { set } from '@ember/object';` - `Ember.set`
* [x] `import { setProperties } from '@ember/object';` - `Ember.setProperties`
* [x] `import { trySet } from '@ember/object';` - `Ember.trySet`
* [x] `import { dependentKeyCompat } from '@ember/object/compat';` - `Ember._dependentKeyCompat`
* [x] `import ComputedProperty from '@ember/object/computed';` - `Ember.ComputedProperty`
* [x] `import { alias } from '@ember/object/computed';` - `Ember.computed.alias`
* [x] `import { and } from '@ember/object/computed';` - `Ember.computed.and`
* [x] `import { bool } from '@ember/object/computed';` - `Ember.computed.bool`
* [x] `import { collect } from '@ember/object/computed';` - `Ember.computed.collect`
* [x] `import { deprecatingAlias } from '@ember/object/computed';` - `Ember.computed.deprecatingAlias`
* [x] `import { empty } from '@ember/object/computed';` - `Ember.computed.empty`
* [x] `import { equal } from '@ember/object/computed';` - `Ember.computed.equal`
* [x] `import { expandProperties } from '@ember/object/computed';` - `Ember.expandProperties`
* [x] `import { filter } from '@ember/object/computed';` - `Ember.computed.filter`
* [x] `import { filterBy } from '@ember/object/computed';` - `Ember.computed.filterBy`
* [ ] ~`import { filterProperty } from '@ember/object/computed';` - `Ember.computed.filterProperty`~
* [x] `import { gt } from '@ember/object/computed';` - `Ember.computed.gt`
* [x] `import { gte } from '@ember/object/computed';` - `Ember.computed.gte`
* [x] `import { intersect } from '@ember/object/computed';` - `Ember.computed.intersect`
* [x] `import { lt } from '@ember/object/computed';` - `Ember.computed.lt`
* [x] `import { lte } from '@ember/object/computed';` - `Ember.computed.lte`
* [x] `import { map } from '@ember/object/computed';` - `Ember.computed.map`
* [x] `import { mapBy } from '@ember/object/computed';` - `Ember.computed.mapBy`
* [ ] ~`import { mapProperty } from '@ember/object/computed';` - `Ember.computed.mapProperty`~
* [x] `import { match } from '@ember/object/computed';` - `Ember.computed.match`
* [x] `import { max } from '@ember/object/computed';` - `Ember.computed.max`
* [x] `import { min } from '@ember/object/computed';` - `Ember.computed.min`
* [x] `import { none } from '@ember/object/computed';` - `Ember.computed.none`
* [x] `import { not } from '@ember/object/computed';` - `Ember.computed.not`
* [x] `import { notEmpty } from '@ember/object/computed';` - `Ember.computed.notEmpty`
* [x] `import { oneWay } from '@ember/object/computed';` - `Ember.computed.oneWay`
* [x] `import { or } from '@ember/object/computed';` - `Ember.computed.or`
* [x] `import { readOnly } from '@ember/object/computed';` - `Ember.computed.readOnly`
* [x] `import { reads } from '@ember/object/computed';` - `Ember.computed.reads`
* [x] `import { setDiff } from '@ember/object/computed';` - `Ember.computed.setDiff`
* [x] `import { sort } from '@ember/object/computed';` - `Ember.computed.sort`
* [x] `import { sum } from '@ember/object/computed';` - `Ember.computed.sum`
* [x] `import { union } from '@ember/object/computed';` - `Ember.computed.union`
* [x] `import { uniq } from '@ember/object/computed';` - `Ember.computed.uniq`
* [x] `import { uniqBy } from '@ember/object/computed';` - `Ember.computed.uniqBy`
* [x] `import CoreObject from '@ember/object/core';` - `Ember.CoreObject`
* [x] `import Evented from '@ember/object/evented';` - `Ember.Evented`
* [x] `import { on } from '@ember/object/evented';` - `Ember.on`
* [x] `import { addListener } from '@ember/object/events';` - `Ember.addListener`
* [x] `import { removeListener } from '@ember/object/events';` - `Ember.removeListener`
* [x] `import { sendEvent } from '@ember/object/events';` - `Ember.sendEvent`
* [x] `import { cacheFor } from '@ember/object/internals';` - `Ember.cacheFor`
* [ ] ~`import { copy } from '@ember/object/internals';` - `Ember.copy`~
* [x] `import { guidFor } from '@ember/object/internals';` - `Ember.guidFor`
* [x] `import Mixin from '@ember/object/mixin';` - `Ember.Mixin`
* [x] `import Observable from '@ember/object/observable';` - `Ember.Observable`
* [x] `import { addObserver } from '@ember/object/observers';` - `Ember.addObserver`
* [x] `import { removeObserver } from '@ember/object/observers';` - `Ember.removeObserver`
* [x] **UNTESTED** `import PromiseProxyMixin from '@ember/object/promise-proxy-mixin';` - `Ember.PromiseProxyMixin`
* [x] **UNTESTED INTERNALLY**`import ObjectProxy from '@ember/object/proxy';` - `Ember.ObjectProxy`

### `@ember/polyfills` (PR: #19947)
* [x] `import { assign } from '@ember/polyfills';` - `Ember.assign`
* [ ] ~`import { create } from '@ember/polyfills';` - `Ember.create`~
* [x] **UNTESTED** `import { hasPropertyAccessors } from '@ember/polyfills';` - `Ember.platform.hasPropertyAccessors`
* [ ] ~`import { keys } from '@ember/polyfills';` - `Ember.keys`~
* [ ] ~`import { merge } from '@ember/polyfills';` - `Ember.merge`~

### `@ember/routing`
* [x] `import { LinkTo } from '@ember/routing';` - `Ember.LinkComponent`
* [x] `import AutoLocation from '@ember/routing/auto-location';` - `Ember.AutoLocation`
* [x] `import HashLocation from '@ember/routing/hash-location';` - `Ember.HashLocation`
* [x] `import HistoryLocation from '@ember/routing/history-location';` - `Ember.HistoryLocation`
* [ ] ~`import LinkComponent from '@ember/routing/link-component';` - `Ember.LinkComponent`~
* [x] `import Location from '@ember/routing/location';` - `Ember.Location`
* [x] `import NoneLocation from '@ember/routing/none-location';` - `Ember.NoneLocation`
* [x] `import Route from '@ember/routing/route';` - `Ember.Route`
* [x] `import EmberRouter from '@ember/routing/router';` - `Ember.Router`
* [x] `import RouterService from '@ember/routing/router-service';` - `Ember.RouterService`

### `@ember/runloop` (PR: #19947)
* [ ] `import { _backburner } from '@ember/runloop';` - `Ember.run.backburner`
* [ ] `import { _cancelTimers } from '@ember/runloop';` - `Ember.run.cancelTimers`
* [ ] `import { _getCurrentRunLoop } from '@ember/runloop';` - `Ember.run._getCurrentRunLoop`
* [ ] `import { _hasScheduledTimers } from '@ember/runloop';` - `Ember.run.hasScheduledTimers`
* [x] `import { begin } from '@ember/runloop';` - `Ember.run.begin`
* [x] `import { bind } from '@ember/runloop';` - `Ember.run.bind`
* [x] `import { cancel } from '@ember/runloop';` - `Ember.run.cancel`
* [x] `import { debounce } from '@ember/runloop';` - `Ember.run.debounce`
* [x] `import { end } from '@ember/runloop';` - `Ember.run.end`
* [x] `import { join } from '@ember/runloop';` - `Ember.run.join`
* [x] `import { later } from '@ember/runloop';` - `Ember.run.later`
* [x] `import { next } from '@ember/runloop';` - `Ember.run.next`
* [x] `import { once } from '@ember/runloop';` - `Ember.run.once`
* [x] `import { run } from '@ember/runloop';` - `Ember.run`
* [x] `import { schedule } from '@ember/runloop';` - `Ember.run.schedule`
* [x] `import { scheduleOnce } from '@ember/runloop';` - `Ember.run.scheduleOnce`
* [x] `import { throttle } from '@ember/runloop';` - `Ember.run.throttle`

### `@ember/service`
* [x] `import Service from '@ember/service';` - `Ember.Service`
* [x] `import { inject } from '@ember/service';` - `Ember.inject.service`
* [x] `import { service } from '@ember/service';`

### `@ember/template` (PR: #19947)
* [x] `import { htmlSafe } from '@ember/template';` - `Ember.String.htmlSafe`
* [x] `import { isHTMLSafe } from '@ember/template';` - `Ember.String.isHTMLSafe`

### `@ember/template-compilation` **NO EXTERNAL TYPES**
* [ ] `import { compileTemplate } from '@ember/template-compilation';` - `Ember.HTMLBars.compile`
* [ ] `import { precompileTemplate } from '@ember/template-compilation';` - `Ember.HTMLBars.precompile`

### `@ember/template-factory` **NO EXTERNAL TYPES**
* [ ] `import { createTemplateFactory } from '@ember/template-factory';` - `Ember.HTMLBars.template`

### `@ember/utils` (PR: #19947)
* [x] `import { compare } from '@ember/utils';` - `Ember.compare`
* [x] `import { isBlank } from '@ember/utils';` - `Ember.isBlank`
* [x] `import { isEmpty } from '@ember/utils';` - `Ember.isEmpty`
* [x] `import { isEqual } from '@ember/utils';` - `Ember.isEqual`
* [x] `import { isNone } from '@ember/utils';` - `Ember.isNone`
* [x] `import { isPresent } from '@ember/utils';` - `Ember.isPresent`
* [ ] ~`import { tryInvoke } from '@ember/utils';` - `Ember.tryInvoke`~
* [x] `import { typeOf } from '@ember/utils';` - `Ember.typeOf`

### `@ember/version` (PR: #19947)
* [x] `import { VERSION } from '@ember/version';` - `Ember.VERSION`

</details>

## Packages to stabilize

This is a list of all types exported by DT. Items should be marked here when they've been compared to the DT types and validated as correct. Note that the DT types could be incorrect in some cases.

- [x] `ember` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/application` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/application/instance` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/array` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/array/mutable` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/array/proxy` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/component` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/component/helper` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/component/template-only` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/controller` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/debug` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/debug/container-debug-adapter` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/debug/data-adapter` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/destroyable` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/engine` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/engine/instance` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/error` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/helper` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/modifier` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/compat` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/computed` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/core` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/evented` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/events` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/internals` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/mixin` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/observable` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/observers` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/promise-proxy-mixin` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/object/proxy` https://github.com/emberjs/ember.js/pull/20449
- [-] `@ember/ordered-set`
- [x] `@ember/owner`: https://github.com/emberjs/ember.js/pull/20288
- [-] `@ember/polyfills` – removed in Ember 5.0
- [x] `@ember/routing` https://github.com/emberjs/ember.js/pull/20449
- [-] `@ember/routing/auto-location` – removed in Ember 5.0
- [x] `@ember/routing/hash-location` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/routing/history-location` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/routing/none-location` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/routing/route` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/routing/router-service` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/routing/router` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/runloop` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/service` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/string` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/template` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/test` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/test/adapter` https://github.com/emberjs/ember.js/pull/20449
- [x] `@ember/utils` https://github.com/emberjs/ember.js/pull/20449

### PR DESCRIPTION
**Pull request by chriskrycho**
🎉🎉🎉🎉🎉

***This PR does exactly what it says on the tin: publishes stable types from Ember's own TypeScript source code!***

🎉🎉🎉🎉🎉

We started working in earnest on writing types for Ember over 6 years ago, in early 2017; we started down the official "Road to TypeScript" 2 full years ago, in early 2021; and I started work on the mechanics we use to publish these types last year. This is one of the last major blockers for implementation for emberjs/rfcs#800, allowing us to have full TypeScript support and it'll make a nice headliner feature for Ember 5.1!

## SemVer

**Note on SemVer:** assuming these types are included in the Ember 5.1 release, we are not *thereby* committed to treating them as stable. They should be treated as stable once we start opting users into them automatically, by including the appropriate imports in the blueprints. That said, we should *plan* to do exactly that, and follow our normal SemVer commitment, as elaborated in [the spec we wrote and adopted for TS types](https://www.semver-ts.org).

## Implementation details

Mechanically:

- Makes a number of small but important updates to the types publishing script, to enable it to handle more edge cases.
- Fixes a number of type definitions internally, when and where the type tests exposed problems with them.
- Adds a number of new types to Ember's internals (and at least "intimate" APIs) which allow it to properly integrate with Glint, using the same mechanics as we used in the DT and preview types packages.
- Removes the `preview` types package entirely, since it is now defunct.
- Updates the type tests themselves to match the implementation where the preview types (previously tested against) were wrong.

Notes on the implementation/for reviewers:

1. There are many commits, and the number of changes here is not small, but it should be comprehensible by working through individual commits, because I have done my very best to make each commit atomic and meaningful… but there are some caveats to that. Because this was an iterative process, there are a few cases where I made a change and then came back to it later in a significantly different way; the key one of these to note is around the `.create()` and `.extend()` types, where the final output is somewhat different than you might think from the initial commits addressing it.

2. I opted to strike a balance between *internal utility* and *external maintainability* for the types we use for the classic class system. In most cases, this had a minimal impact on our internal usage. Notably, it means we do provide *slightly* more support than we originally promised in the RFC. On balance, I think this is fine; in general, we will still strongly nudge people toward native classes because the support is finicky at best. This was the only way to preserve our reliance on some of those features internally without writing unsafe casts for the external types, so it seemed best.

3. I would have liked to break this up into smaller pieces at a time, and indeed did so with what chunks I could last year. However, after taking four or five different passes at that, I realized that there is far, far too much coupling for that to be tractable as things stand. We made some good progress on the internal coupling last year, and I expect that it will become increasingly tractable over 5.0 and, if we deprecate the Classic mechanics during v5, after their removal at v6.0. For now, however, things are too messy, so the only way to ship *any* of the remaining types was to ship *all* of them.

4. This publishes *all* the types: internal and external alike. It would be nice to have a more robust module structure which does not require us to do that, but that is a significantly larger lift. This at least affords us the ability to publish our types!

## Thanks

- LinkedIn sponsored this work—I've spent about a quarter of my time over the last 12 months preparing for this and then delivering on it.
- @dfreeman helped me repeatedly when stuck on extremely thorny type issues, and he and @jamescdavis have helped drive the TS-in-Ember project forward alongside me end to end.
- @cafreeman and @rwjblue provided great sounding boards, feedback, and input on details along the way.

### CODE REVIEW

File: packages/@ember/-internals/utility-types/index.ts [hunk 1]
```diff
@@ -19,3 +19,12 @@ export type MethodParams<T, M extends MethodNamesOf<T>> = Parameters<MethodsOf<T
 export type MethodReturns<T, M extends MethodNamesOf<T>> = ReturnType<MethodsOf<T>[M]>;
 
 export type OmitFirst<F> = F extends [any, ...infer R] ? R : [];
+
+// A way of representing non-user-constructible types. You can conveniently use
+// this by doing `interface Type extends Data<'some-type-name'> { ... }` for
```
Comment (dfreeman) [inline packages/@ember/-internals/utility-types/index.ts:24]:
```suggestion
// this by doing `interface Type extends Opaque<'some-type-name'> { ... }` for
```

File: type-tests/@ember/component-test/helper.ts [hunk 2]
```diff
@@ -153,17 +133,22 @@ const badReturnSig = helper<DemoSig>(([i18nizer], { name, age }) =>
...
 
+// Check that generics are accepted. (We cannot meaningfully test for how they
+// are *handled* because it uses items we do not want to expose in public API.)
 const pair = helper(<T>([item]: [T]): [T, T] => [item, item]);
-expectTypeOf(pair).toEqualTypeOf<
-  abstract new <T>() => FunctionBasedHelperInstance<{
-    Args: { Positional: [T]; Named: EmptyObject };
-    Return: [T, T];
-  }>
->();
```
Comment (dfreeman) [inline type-tests/@ember/component-test/helper.ts:169]: What will TS produce when emitting declarations for a helper like this if `FunctionBasedHelperInstance` isn't exported?

**Comment by dfreeman** [comment]
I left a few comments/questions, and am hoping to get the Glint monorepo typechecking against a build from this branch tomorrow to make sure everything works cleanly there, but overall this looks fantastic! 🎉

File: packages/@ember/-internals/glimmer/lib/components/input.ts [hunk 3]
```diff
@@ -270,4 +270,6 @@ class Input extends AbstractInput {
   }
 }
 
-export default opaquify(Input, InputTemplate);
+const Input = opaquify(_Input, InputTemplate);
+type Input = typeof Input;
```
[the remark this answers is not in the document]
Comment (chriskrycho) [reply packages/@ember/-internals/glimmer/lib/components/input.ts:274]: Ah, yes:

https://github.com/emberjs/ember.js/blob/b1ebc7d4abf99b04168e824137d8e2ecd3a329a6/types/preview/@ember/component/index.d.ts#L145-L146

I'll update all of these to match that. Thanks!

File: types/publish.mjs [hunk 4]
```diff
@@ -97,94 +96,136 @@ ${MODULES_PLACEHOLDER}
...
+      output: 'require.d.ts',
+    },
+    {
+      input: 'packages/@ember/service/owner-ext.d.ts',
+      output: '@ember/service/owner-ext.d.ts',
+    },
+    {
+      input: 'packages/@ember/routing/service-ext.d.ts',
+      output: '@ember/routing/service-ext.d.ts',
+    },
```
[the remark this answers is not in the document]
Comment (chriskrycho) [reply types/publish.mjs]: I don't think you're missing anything; I think this is what happens when the authoring of a given piece of infrastructure gets chopped up over the course of many months by massive conflagrations in unrelated areas. 😂 I'll see if I can combine them!

[the remark this answers is not in the document]
Comment (chriskrycho) [reply types/publish.mjs]: Fixed via 30a80963e! Separated and improved them.

File: packages/ember/index.ts [hunk 5]
```diff
@@ -124,264 +157,281 @@ import type * as EmberTemplateCompiler from 'ember-template-compiler';
...
   // ****@ember/canary-features****
-  FEATURES: { isEnabled, ...FEATURES },
+  export const FEATURES = { isEnabled, ...EmberFEATURES };
 
   // ****@ember/component****
-  _Input: Input,
-  Component,
+  export const _Input = EmberInput;
+  export const Component = EmberComponent;
+  export type Component = EmberComponent;
```
[the remark this answers is not in the document]
Comment (chriskrycho) [reply packages/ember/index.ts:264]: Ah, it definitely should!

[the remark this answers is not in the document]
Comment (chriskrycho) [reply packages/ember/index.ts:264]: 4bc5d73f

File: types/publish.mjs [hunk 6]
```diff
@@ -97,94 +96,136 @@ ${MODULES_PLACEHOLDER}
...
 
   process.exit(status === 'success' ? 0 : 1);
 }
 
+/**
+  "Emit" hand-authored `.d.ts` modules for two cases:
+
+  - modules which need to live in a different location in the output than in the
+    input tree, e.g. for the loader, which creates runtime modules at a
+    different location than
```
[the remark this answers is not in the document]
Comment (chriskrycho) [reply types/publish.mjs]: It was! Fixed when I reworked all of this in 22877513.

File: packages/@ember/routing/route.ts [hunk 7]
```diff
@@ -254,14 +248,11 @@ interface Route<T = unknown> extends IRoute<T> {
...
-class Route<T = unknown>
-  extends EmberObject.extend(ActionHandler, Evented)
-  implements IRoute, Evented
-{
+class Route<Model = unknown> extends EmberObject.extend(ActionHandler, Evented) implements IRoute {
   static isRouteFactory = true;
 
-  context = {} as T;
-  declare currentModel: T;
+  context = {} as Model;
```
[the remark this answers is not in the document]
Comment (chriskrycho) [reply packages/@ember/routing/route.ts:254]: It is not. Unfortunately, it is (apparently?) part of the public API of `router.js`'s `Route` type, so we cannot make it private *here* without also making it private *there*, since this is typed as `implements IRoute` (which is the alias for the one for `router.js`). 😩

[the remark this answers is not in the document]
Comment (chriskrycho) [reply packages/@ember/routing/route.ts:259]: Marked with `@internal` as better-than-nothing signal. b7a7a57c

File: packages/@ember/runloop/type-tests/bind.test.ts [hunk 8]
```diff
@@ -135,4 +135,6 @@ expectTypeOf(bind(foo, 'test', 1, true, 'baz')).toEqualTypeOf<() => number | voi
 expectTypeOf(bind(foo, 'test', 1, true, undefined)).toEqualTypeOf<() => number | void>();
 
 // @ts-expect-error Invalid args
+bind(foo, foo.test, 'string');
+// We would like to catch this, but it ends up in the fallback path.
 bind(foo, 'test', 'string');
```
[the remark this answers is not in the document]
Comment (chriskrycho) [reply packages/@ember/runloop/type-tests/bind.test.ts]: Good call. Done in 19b31a35e.

File: type-tests/@ember/component-test/helper.ts (hunk 2 above)
Comment (chriskrycho) [reply type-tests/@ember/component-test/helper.ts]: Did some poking and confirmed that we do need this type exported. @dfreeman and I spent a bunch of time discussing alternatives, but ultimately landed on "just export this thing and look forward to when the need for it goes away by dropping `helper()` in favor of *just* using plain functions instead"; done in c725cdce7.

**Comment by chriskrycho** [comment]
**Status note:** We have hit a bit of a snag with this (though not an intractable one)!

All of the stuff I built here works—and works correctly—but you can’t actually use it in a consuming app or addon yet, because `ember-source` currently puts `@glimmer/*` as `devDependencies` and pre-builds all of that as part of its build—but that means that those types are unavailable, and they’re inherently part of the types we publish.

(Where possible I have decreased that exposure in public APIs, but it is and will for the foreseeable future be impossible for that exposure to be *zero*. This problem is not going away.)

Net, when you try to use the types published from Ember's source, you end up with an awful lot of these (45 across 28 different files in a fresh addon):

```
node_modules/ember-source/types/stable/ember/index.d.ts:110:55 - error TS2307: Cannot find module '@glimmer/manager' or its corresponding type declarations.

110     const _modifierManagerCapabilities: typeof import('@glimmer/manager').modifierCapabilities;
                                                          ~~~~~~~~~~~~~~~~~~
```

One option here is to update `ember-source` to depend on those as `dependencies` instead of as `devDependencies`, just the way we would for a "normal" addon. This will be a mild bit of “incoherence” in that those dependencies *will* be the ones Ember uses… but the copy in `node_modules` will *not* be the ones you actually get, courtesy of the pre-build. (However, I suspect we may want to do that *anyway* as part of moving toward Ember’s packages being normal packages?)

We don’t yet have the ability to use something like `rollup-plugin-ts` which might pull these dependency in for us "automatically" by way of the types rollup, because we still have too many cycles in our packages (hard blocker) *and* a small number of straight-up lies in the form of code we generate and which has no on-disk representation (more solvable but not trivial).

I am discussing this with other folks on Framework Core and Typed Ember Core, and will update once we agree on a path forward.

**Review by dfreeman — APPROVED** [review]

**Comment by chriskrycho** [comment]
After discussing with folks on Friday, we're going to go ahead and merge this. It *should* be free of issues, but we will keep a close eye out as we move forward. This should be present in Ember 5.1! 🎉
